x-pack/plugin/mapper-version/src/main/java/org/elasticsearch/xpack/versionfield/Version.java

```diff
@@ -17,7 +19,7 @@
  * TODO(stu): implement {@code Comparable<Version>} based on {@code VersionEncoder#prefixDigitGroupsWithLength(String, BytesRefBuilder)}
  * See: https://github.com/elastic/elasticsearch/issues/82287
  */
-public class Version implements ToXContent {
+public class Version implements ToXContent, BytesRefProducer, Comparable<Version> {
     protected String version;
 
     public Version(String version) {
```

Comment:
I wonder if we should have another constructor that takes the encoded version, this String version is mostly for users.

Reply:
Do you mean a `protected Version(String version, BytesRef ref)` for internal use only, so that we don't have to recalculate the BytesRef? I guess having a constructor with BytesRef would have the same problem, ie. we would have to decode it for the string value.

Reply:
From a quick look I think we can have a ctor that only takes the BytesRef and internally decodes it to a String only once, then use one for the xContent/toString part and the other for comparison. `Version` only seems to be created from `VersionStringDocValuesField` where we can also directly get the bytes, in fact currently we do the decoding step there once in `getInternal`, that could happen in the ctor instead.

Reply:
We still need the `Version(String)` constructor is [whitelisted](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/mapper-version/src/main/resources/org/elasticsearch/xpack/versionfield/org.elasticsearch.xpack.versionfield.txt#L15) for painless users.  Users have to have the ability to provide a `Version` as a default value when fetching version fields, as in this [test](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/mapper-version/src/yamlRestTest/resources/rest-api-spec/test/50_script_values.yml#L40).

Reply:
Yes, I'm keeping both (see last commit)